pennylane/data/base/dataset.py

```diff
@@ -0,0 +1,265 @@
+        default=default,
+        default_factory=default_factory,
+    )
+
+
+@dataclass_transform(
+    order_default=False, eq_default=False, field_specifiers=(attribute,), kw_only_default=True
+)
+class _DatasetMeta(type):
+    """This is a metaclass that tells the type system that
+    ``DatasetBase`` behaves like a dataclass. See:
+    https://peps.python.org/pep-0681/
+    """
+
+
+class Dataset(MapperMixin, metaclass=_DatasetMeta):
+    """
+    Base class for Datasets.
+
+    Attributes:
+        fields: A mapping of attribute names to their ``Attribute`` information. Note that
+            this contains attributes declared on the class, not attributes added to
+            an instance. Use ``attrs`` to view all attributes on an instance.
+        bind: The Zarr group that contains this dataset.
+    """
+
+    type_id = "dataset"
+
+    fields: ClassVar[typing.Mapping[str, Attribute]] = MappingProxyType({})
+
+    bind: InitVar[Optional[ZarrGroup]] = attribute(default=None, kw_only=False)  # type: ignore
+
+    def __init__(
+        self,
+        bind: Optional[ZarrGroup] = None,
+        **attrs: Any,
+    ):
+        """
+        Load a dataset from a Zarr Group or initialize a new Dataset.
+
+        Args:
+            bind: The Zarr group, or path to zarr file, that will contain this dataset.
+                If None, the dataset will be stored in memory. Any attributes that
+                already exist in ``bind`` will be loaded into this dataset.
+            **attrs: Attributes to add to this dataset.
+        """
+        if bind is None:
+            self._bind = zarr.group()
+            self._bind.attrs["type_id"] = self.type_id
+        else:
+            self._bind = bind
+
+        self._validate_arguments(attrs)
+        for name, attr in attrs.items():
+            setattr(self, name, attr)
+
+    @property
+    def bind(self) -> ZarrGroup:
+        """The Zarr group that contains this dataset's attributes."""
+        return self._bind
+
+    @property
+    def attrs(self) -> typing.Mapping[str, AttributeType]:
+        """Returns all attributes of this Dataset."""
+        return self._mapper.view()
+
+    def list_attributes(self) -> List[str]:
+        """Returns a list of this dataset's attributes."""
+        return list(self.attrs.keys())
+
+    def read(
+        self,
+        filepath: Union[str, Path],
+        assign_to: Optional[str] = None,
+        *,
+        overwrite_attrs: bool = False,
+    ):
+        """Load dataset from Zarr file at filepath. Can also accept an S3 URL.
+
+        Args:
+            filepath: Path to file containing dataset
+            assign_to: Attribute name to which the contents of the file should be assigned.
+                If this is ``None`` (the default value), the file's attributes will be assigned
+                to the current dataset
+            overwrite_attrs: Whether to overwrite attributes that already exist in this
+                dataset.
```

Comment:
When trying to overwrite an existing attribute Zarr outputs the following error:
`CopyError: an object 'attribute_name' already exists in destination '/'`

Could be good to catch this and and return our own error to let users what is happening and suggest setting `overwrite_attrs=True`

Reply:
good point, I'll add a TODO for this